models/concerns/validate_data.rb

```diff
@@ -140,9 +140,108 @@ def check_shipments_specificities
     }
   end
 
+  def calculate_day_availabilities(vehicles, timewindows)
+    vehicles_days = timewindows.collect{ |timewindow|
+      if timewindow.empty?
+        []
+      else
+        days = timewindow.flat_map{ |tw| tw[:day_index] }
+        days.compact!
+        days.uniq
+      end
+    }.delete_if(&:empty?)
+
+    vehicles_unavailable_indices = vehicles.collect{ |v| v[:unavailable_work_day_indices] }
+    vehicles_unavailable_indices.compact!
+
+    [vehicles_days, vehicles_unavailable_indices]
+  end
+
+  def check_vehicle_trips_stores_consistency(relation_vehicles)
+    relation_vehicles.each_with_index{ |vehicle_trip, v_i|
+      case v_i
+      when 0
+        unless vehicle_trip[:end_point_id]
+          raise OptimizerWrapper::DiscordantProblemError.new('First trip should at least have an end point id')
+        end
+      when relation_vehicles.size - 1
+        unless vehicle_trip[:start_point_id]
+          raise OptimizerWrapper::DiscordantProblemError.new('Last trip should at least have a start point id')
+        end
+      else
+        unless vehicle_trip[:start_point_id] && vehicle_trip[:end_point_id]
+          raise OptimizerWrapper::DiscordantProblemError.new(
+            'Intermediary trips should have a start and an end point ids'
+          )
+        end
+      end
+
+      next if v_i.zero? ||
+              vehicle_trip[:start_point_id] == relation_vehicles[v_i - 1][:end_point_id] ||
+              (@hash[:points].find{ |pt| pt[:id] == vehicle_trip[:start_point_id] }[:location] ==
+                @hash[:points].find{ |pt| pt[:id] == relation_vehicles[v_i - 1][:end_point_id] }[:location])
+
+      raise OptimizerWrapper::DiscordantProblemError.new('One trip should start where the previous trip ended')
+    }
+  end
+
+  def check_trip_timewindows_consistency(relation_vehicles)
+    vehicles_timewindows =
+      relation_vehicles.collect{ |v| v[:timewindow] ? [v[:timewindow]] : v[:sequence_timewindows].to_a }
+
+    return if vehicles_timewindows.all?(&:empty?)
+
+    week_days, unavailable_indices = calculate_day_availabilities(relation_vehicles, vehicles_timewindows)
+    if week_days.uniq.size > 1 || unavailable_indices.uniq.size > 1
+      raise OptimizerWrapper::DiscordantProblemError.new(
+        'Vehicles in vehicle_trips relation should have the same available days'
+      )
+    end
+
+    vehicles_timewindows.each_with_index{ |v_tw, v_i|
+      next if v_i.zero?
+
+      v_tw.each{ |tw|
+        day = tw[:day_index]
```

Comment:
what if a vehicle has a timewindow {start: 0:00, end:8:00, day_index: 1} and the previous one in the trip has {start: 16:00, end: 23:45, day_index:0} with a lapse of one hour ?

Reply:
For now, vehicles trip should take place at the exact same day, so this case is not supported yet.

Reply:
The above check verify that all vehicles have the same days, not only one and exact same day ?